src/routers/alpha-router/functions/get-candidate-pools.ts

```diff
@@ -197,16 +201,140 @@ class SubcategorySelectionPools<SubgraphPool> {
   constructor(
     public pools: SubgraphPool[],
     public readonly poolsNeeded: number
-  ) {}
+  ) {
+  }
 
   public hasEnoughPools(): boolean {
     return this.pools.length >= this.poolsNeeded;
   }
 }
 
+export type CrossedLiquidityCandidatePools = {
+  v2Pools: V2SubgraphPool[];
+  v3Pools: V3SubgraphPool[];
+}
+
+export async function getMixedCrossLiquidityCandidatePools({
+  tokenIn,
+  tokenOut,
+  routingConfig,
+  v2SubgraphProvider,
+  v3SubgraphProvider,
+  v2Candidates,
+  v3Candidates,
+}: MixedCrossLiquidityCandidatePoolsParams): Promise<CrossedLiquidityCandidatePools> {
+  const blockNumber = routingConfig.blockNumber;
+
+  const v2Pools = (await v2SubgraphProvider.getPools(tokenIn, tokenOut, {
+    blockNumber,
+  })).sort((a, b) => b.reserve - a.reserve);
+  const v3Pools = (await v3SubgraphProvider.getPools(tokenIn, tokenOut, {
+    blockNumber,
+  })).sort((a, b) => b.tvlUSD - a.tvlUSD);
+
+  const tokenInAddress = tokenIn.address.toLowerCase();
+  const tokenOutAddress = tokenOut.address.toLowerCase();
+
+  const v2SelectedPools: {
+    forTokenIn?: V2SubgraphPool;
+    forTokenOut?: V2SubgraphPool;
+  } = {};
+  const previouslySelectedV2PoolIds = new Set(v2Candidates?.subgraphPools.map(pool => pool.id) ?? []);
+
+  const crossTokenAgainstV3TokenOut = v3Candidates?.candidatePools.selections.topByTVLUsingTokenOut[0]?.token0.id.toLowerCase() === tokenOutAddress ?
+    v3Candidates?.candidatePools.selections.topByTVLUsingTokenOut[0]?.token1.id.toLowerCase() :
+    v3Candidates?.candidatePools.selections.topByTVLUsingTokenOut[0]?.token0.id.toLowerCase();
+  const crossTokenAgainstV3TokenIn = v3Candidates?.candidatePools.selections.topByTVLUsingTokenIn[0]?.token0.id.toLowerCase() === tokenInAddress ?
+    v3Candidates?.candidatePools.selections.topByTVLUsingTokenIn[0]?.token1.id.toLowerCase() :
+    v3Candidates?.candidatePools.selections.topByTVLUsingTokenIn[0]?.token0.id.toLowerCase();
```

Comment:
Find the token (hop) for either tokenIn or tokenOut, we will pull these from the `topByTvl` criteria on the V3 candidates, then we will look for the "other side" of the route among the v2Pools